### PR TITLE
Add return units for models with an amplitude parameter

### PR DESCRIPTION
**Pull request by lpsinger**
### Description
It's useful to be able to introspect the return units of models. There is a `return_units` property that is part of the public Model class API, but most models do not define it.

This patch implements `return_units` for the very common case that a model consists of a dimensionless function times a dimensionful amplitude parameter.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Need change log and tests. Thanks!

**Comment by lpsinger** [comment]
Done.

**Comment by pllim** [comment]
Seems like failures are real.

`E       astropy.units.core.UnitConversionError: '1 / Hz' (time) and 'km / s' (speed/velocity) are not convertible`

**Comment by lpsinger** [comment]
I don't see what failed in the readthedocs build.

**Comment by pllim** [comment]
Perhaps a transient server issue. I restarted the RTD job. 🤞

**Comment by pllim** [comment]
Hrmmm let's try again after #12929 is resolved.

**Comment by eerovaher** [comment]
There's a bunch of warnings like this:
```
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/12926/lib/python3.9/site-packages/astropy/modeling/functional_models.py:docstring of astropy.modeling.functional_models.AiryDisk2D:1: WARNING: py:class reference target not found: astropy.modeling.mixins.AmplitudeReturnUnitMixin
```
It looks like `docs/modeling/reference_api.rst` needs to be updated.

**Review by WilliamJamieson — COMMENTED** [review]
I am currently, a bit uncomfortable with these changes. @nden and @perrygreenfield what are your opinions?

File: astropy/modeling/mixins.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+"""
+
+
+class AmplitudeReturnUnitMixin:
+
+    @property
+    def return_units(self):
+        if self.amplitude.unit is None:
+            return None
+        return dict.fromkeys(self.outputs, self.amplitude.unit)
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply astropy/modeling/mixins.py:14]: While use of mixins is not uncommon in Python, my sense (perhaps mistaken) is that a mixin should try to be completely orthogonal to any other class it is mixed with. And that would include overriding that class's methods. So I agree with that aspect of this comment. I also wonder if it shouldn't be made an optional part of the original class.

**Comment by pllim** [comment]
Looks like there are ongoing discussions and also conflicts, so moving the milestone.

**Pull request closed without merging**